torch/csrc/autograd/functions/special.cpp

```diff
@@ -129,45 +127,46 @@ bool Eval::trySimpleEval(const variable_list& inputs, const variable_list& outpu
-  if (static_cast<std::size_t>(grad_fn->num_inputs) != outputs.size()) return false;
+  if (static_cast<std::size_t>(grad_fn->num_inputs()) >= max_outputs) return false;
+  if (static_cast<std::size_t>(grad_fn->num_inputs()) != outputs.size()) return false;
 
   // Check that all outputs have the same grad_fn and cover all its inputs
   bitset_type output_nrs = 0;
-  bitset_type expected_bitset = ((1 << grad_fn->num_inputs) - 1);
+  bitset_type expected_bitset = ((1 << grad_fn->num_inputs()) - 1);
   for (auto & output : outputs) {
     if (output.grad_fn() != grad_fn) return false;
     output_nrs |= (1 << output.output_nr());
   }
   if (output_nrs != expected_bitset) return false;
 
-  // Check that grad_fn->next_functions matches the inputs exactly
+  // Check that grad_fn's next_edges match the inputs exactly.
   auto num_inputs = inputs.size();
-  auto& grad_next_fns = grad_fn->next_functions;
-  if (num_inputs != grad_next_fns.size()) return false;
+  if (num_inputs != grad_fn->num_outputs()) return false;
   for (std::size_t i = 0; i < num_inputs; ++i) {
-    // Unfortunately, null edge pruning (see Note [Null-edge pruning]) applies to
-    // autograd functions which would otherwise be eligible for the SimpleEval
-    // optimization.  This makes everything more complicated, so for now we just don't
-    // attempt the optimization in this case.  To fix it properly,
-    // we'd need to filter grad_next_fns and outputs of apply in Eval::apply.
-    // The check below tests if null edge pruning occurred.
-    if (!inputs[i].defined() || !grad_next_fns[i].is_valid()) return false;
-    if (grad_next_fns[i] != inputs[i].gradient_edge()) return false;
+    const auto& next_grad_edge = grad_fn->next_edge(i);
+    // Unfortunately, null edge pruning (see Note [Null-edge pruning]) applies
+    // to autograd functions which would otherwise be eligible for the
+    // SimpleEval optimization.  This makes everything more complicated, so for
+    // now we just don't attempt the optimization in this case.  To fix it
+    // properly, we'd need to filter grad_fn's output edges and outputs of
+    // apply in Eval::apply. The check below tests if null edge pruning
+    // occurred.
+    if (!inputs[i].defined() || !next_grad_edge.is_valid()) return false;
+    if (next_grad_edge != inputs[i].gradient_edge()) return false;
   }
 
   // Success! We still need to set up placeholders for next stages and to drop
   // references to the graph.
-  std::swap(next_functions, grad_next_fns);
-  grad_next_fns.reserve(num_inputs);
+  next_edges_.reserve(num_inputs);
+  grad_fn->swap_next_edges(next_edges_);
   placeholders.reserve(num_inputs);
-  for (std::size_t i = 0; i < num_inputs; ++i) {
-    auto placeholder = std::make_shared<EvalOutput>(next_functions[i]);
-    grad_next_fns.emplace_back(placeholder, 0);
+  for (const auto& input : next_edges_) {
+    auto placeholder = std::make_shared<EvalOutput>(input);
+    grad_fn->add_next_edge({placeholder, 0});
```

Comment:
I can call it `create_gradient_edge`. I had that name first but changed it to `add` because of the overload I added that takes an `Edge` (since it wasn't "creating" an Edge). Since I'm bringing back `set_gradient_edge`, the free function can `create` and edge. At least it'll be `create` instead of `add`